File: sdk-java/src/main/java/io/littlehorse/sdk/worker/internal/LHServerConnection.java
package io.littlehorse.sdk.worker.internal;

import io.grpc.stub.StreamObserver;
import io.littlehorse.sdk.common.LHLibUtil;
import io.littlehorse.sdk.common.proto.LHHostInfo;
import io.littlehorse.sdk.common.proto.LittleHorseGrpc.LittleHorseStub;
import io.littlehorse.sdk.common.proto.PollTaskRequest;
import io.littlehorse.sdk.common.proto.PollTaskResponse;
import io.littlehorse.sdk.common.proto.ScheduledTask;
import java.io.Closeable;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class LHServerConnection implements Closeable, StreamObserver<PollTaskResponse> {

    private LHServerConnectionManager manager;
    private LHHostInfo host;

    private boolean stillRunning;
    private StreamObserver<PollTaskRequest> pollClient;
    private LittleHorseStub stub;

    public LHServerConnection(LHServerConnectionManager manager, LHHostInfo host) {
        stillRunning = true;
        this.manager = manager;
        this.host = host;

        this.stub = manager.config.getAsyncStub(host.getHost(), host.getPort());

        this.pollClient = this.stub.pollTask(this);

        askForMoreWork();
    }

    @Override
    public void onCompleted() {
        // This shouldn't happen.
        log.error("Unexpected call to onCompleted() in the Server Connection.");
        this.stillRunning = false;
        manager.onConnectionClosed(this);
    }

    @Override
    public void onError(Throwable t) {
        log.error("Unexpected error from server", t);
        this.stillRunning = false;
        manager.onConnectionClosed(this);
    }

    @Override
    public void onNext(PollTaskResponse taskToDo) {
        if (taskToDo.hasResult()) {
            ScheduledTask scheduledTask = taskToDo.getResult();
            String wfRunId = LHLibUtil.getWfRunId(scheduledTask.getSource()).getId();
            log.debug("Received task schedule request for wfRun {}", wfRunId);

            manager.submitTaskForExecution(scheduledTask, this.stub);

            log.debug("Scheduled task on threadpool for wfRun {}", wfRunId);
        } else {
            log.error("Didn't successfully claim task, likely due to server restart.");
        }

        if (stillRunning) {
            askForMoreWork();
        } else {
            // This may cause issues when there's multiple threads, eg closing an
            // already closed stream observer.
            pollClient.onCompleted();
        }
    }

    public LHHostInfo getHostInfo() {
        return host;
    }

    public boolean isSameAs(LHHostInfo other) {
        return (this.host.getHost().equals(other.getHost()) && this.host.getPort() == other.getPort());
    }

    private void askForMoreWork() {
        log.debug("Asking for more work on {}:{}", host.getHost(), host.getPort());
        pollClient.onNext(PollTaskRequest.newBuilder()
                .setClientId(manager.config.getTaskWorkerId())
                .setTaskDefId(manager.taskDef.getId())
                .setTaskWorkerVersion(manager.config.getTaskWorkerVersion())
                .build());
    }

    public void close() {
        stillRunning = false;

        // Let the LH server know we're done
        pollClient.onCompleted();
    }
}

File: sdk-dotnet/LittleHorseSDK/LittleHorse/Worker/Internal/LHServerConnection.cs
using Grpc.Core;
using LittleHorse.Worker.Internal.Helpers;
using LittleHorse.Common.Proto;
using Microsoft.Extensions.Logging;

namespace LittleHorse.Worker.Internal
{
    public class LHServerConnection<T> : IDisposable
    {
        private LHServerConnectionManager<T> _connectionManager;
        private LHHostInfo _hostInfo;
        private bool _running;
        private LittleHorse.Common.Proto.LittleHorse.LittleHorseClient _client;
        private AsyncDuplexStreamingCall<PollTaskRequest, PollTaskResponse> _call;
        private ILogger? _logger;

        public LHHostInfo HostInfo { get { return _hostInfo; } }

        public LHServerConnection(LHServerConnectionManager<T> connectionManager, LHHostInfo hostInfo, ILogger? logger = null)
        {
            _connectionManager = connectionManager;
            _hostInfo = hostInfo;
            _logger = logger;
            _client = _connectionManager.Config.GetGrcpClientInstance(_hostInfo.Host, _hostInfo.Port);
            _call = _client.PollTask();
        }

        public void Connect()
        {
            _running = true;
            Task.Run(RequestMoreWorkAsync);
        }

        private async Task RequestMoreWorkAsync()
        {


            var request = new PollTaskRequest()
            {
                ClientId = _connectionManager.Config.WorkerId,
                TaskDefId = _connectionManager.TaskDef.Id,
                TaskWorkerVersion = _connectionManager.Config.TaskWorkerVersion
            };

            var readTask = Task.Run(async () =>
             {
                 await foreach (var taskToDo in _call.ResponseStream.ReadAllAsync())
                 {
                     if (taskToDo.Result != null)
                     {
                         var scheduledTask = taskToDo.Result;
                         var wFRunId = LHWorkerHelper.GetWFRunId(scheduledTask.Source);
                         _logger?.LogDebug($"Received task schedule request for wfRun {wFRunId}");

                         _connectionManager.SubmitTaskForExecution(scheduledTask, _client);

                         _logger?.LogDebug($"Scheduled task on threadpool for wfRun {wFRunId}");
                     }
                     else
                     {
                         _logger?.LogError($"Didn't successfully claim task, likely due to server restart.");
                     }

                     if (_running)
                     {
                         await _call.RequestStream.WriteAsync(request);
                         _logger?.LogDebug($"Request work on {_hostInfo.Host} : {_hostInfo.Port}");
                     }
                     else
                     {
                         await _call.RequestStream.CompleteAsync();
                     }
                 }
             });

            await _call.RequestStream.WriteAsync(request);
            _logger?.LogDebug($"Request work on {_hostInfo.Host} : {_hostInfo.Port}");
            await readTask;
        }

        public void Dispose()
        {
            _running = false;
        }

        public bool IsSame(LHHostInfo hostInfoToCompare)
        {
            return _hostInfo.Host.Equals(hostInfoToCompare.Host) && _hostInfo.Port == hostInfoToCompare.Port;
        }
    }
}
